### ISSUE TITLE
Handling of encrypted content to be sent to the Continuing and Replay queues

### ISSUE THREAD

**Issue by aspacca** [elastic/elastic-serverless-forwarder#305]
**Describe the enhancement:**

The content of the original inputs of the Forwarder could be encrypted: the decryption happens transparently in the Forwarder, just providing it the ARNs of the KMS Keys at deployment time so that the Forwarder lambda can be granted the proper IAM permissions required.
There can be the case where (part of) the original content is sent to the continuing queue or the replay queue: in this case the content is sent decrypted, lowering the security level of the original content.
We should provide a way to send the original content to the two above queues encrypted: my suggestion is to make the two queues encrypted by default, without the option of opting-out from encryption. We'll generate the KMS Keys at deployment time and configure the proper settings for the two queues. The two queues in the existing deployments will change to be encrypted after the given release.

Existing messages in the queues before making them encrypted will remain unencrypted: it has to be investigated if this requires special handling in the Forwarder code, or is it a blocker at all (can the lambda receive a mix of unencrypted and encrypted messages from the same queue?)

**Comment by aspacca** [comment]
`SqsManagedSseEnabled` property is available for `AWS::SQS::Queue` resource (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sqs-queue.html#cfn-sqs-queue-sqsmanagedsseenabled) in order to enable SSE-SQS encryption (encryption at rest using AWS managed key).

Our CloudFormation templates, both with SAR deployment or direct deployment, do not specify this property

In September/October 2022 newly created queues with the property not specified will default to `true` (public announcement is date 5 Oct 2022: https://aws.amazon.com/blogs/compute/announcing-server-side-encryption-with-amazon-simple-queue-service-managed-encryption-keys-sse-sqs-by-default/, release notes is dated 26 Sept 2022: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-release-notes.html)

Direct deployment was introduced with v1.6.0 release, on 1 Feb 2023: all the direct deployments have the Continuing and the Replay queues encrypted.

SAR deployments before and included v1.3.0 have the two queues unencrypted: according to my test, even if the users updated to a newer version, existing queues without `SqsManagedSseEnabled` property specified will not have SSE-SQS enabled automatically according to the new default value.

For this reason we must expressly specify `SqsManagedSseEnabled: true` property in the CloudFormation template.

Once SSE-SQS is enabled on queues where it was previously disabled, the queues will contain a mix a unencrypted and encrypted messages: according to my test no change in the code is required in order to handle this case. But we must add a specific integration test for this scenario.

Finally there are costs and performance considerations to make.
There are two available encryption method in SQS: the mentioned SSE-SQS and SSE-KMS.
SSE-SQS uses the AWS managed KMS key and has no additional costs. It is currently unknown how many calls to AWS KMS are executed with this method.

SSE-KMS allow the users to provide a specific KMS Key with the `KmsMasterKeyId` property (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sqs-queue.html#cfn-sqs-queue-kmsmasterkeyid), including the AWS managed one used by SSE-SQS.
In this case it is possible to set the `KmsDataKeyReusePeriodSeconds` property (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sqs-queue.html#cfn-sqs-queue-kmsdatakeyreuseperiodseconds), that's the length of time in seconds for which Amazon SQS can reuse a data key to encrypt or decrypt messages before calling AWS KMS again.
Here's the note from AWS documentation:
```
A shorter time period provides better security, but results in more calls to AWS KMS, which might incur charges after Free Tier
```

Setting `KmsDataKeyReusePeriodSeconds` with `SqsManagedSseEnabled` is a noop and the property is ignored.

From the costs point of view, SSE-SQS is the most convenient solution, as it has not additional charges, but we have no visibility on the impact of the number of calls to AWS KMS: ie the impact on performance.

Note that it is possible to set the following:
```
KmsMasterKeyId: alias/aws/sqs # the AWS managed KMS key
KmsDataKeyReusePeriodSeconds: 86400
```

I'm waiting for confirmation if, in this case, considering we use the AWS managed key used by SSE-SQS, this will result in no additional costs as well.

I'm also waiting to know what it is the equivalent value for `KmsDataKeyReusePeriodSeconds`, in terms of number of calls to AWS KMS, when using SSE-SQS compared to SSE-KMS

According to the answer of two open questions above we might or might have not to trade off between costs optimisation and performance optimisation.
If we don't need to trade off, I'd suggest we go with SSE-SQS.
If we need to trade off, I'd suggest to expose a setting to the users whether to use SSE-SQS with no additional charges, or use SSE-KMS with the AWS managed KMS key with the highest value for `KmsDataKeyReusePeriodSeconds` so to optimise performance even if this might incur in additional charges. Potentially we could document the potential additional charges the users can expect.

**Comment by aspacca** [comment]
Current key reuse period for Managed SSE for SQS is set to 24 hours (86400 seconds), so the performance impact for an SQS queue with enabled Managed SSE (SSE-SQS) should be roughly the same as with customer-managed SSE (SSE-KMS).

As final outcome we must expressly specify `SqsManagedSseEnabled: true` property in the CloudFormation template.
